src/cluster_legacy.c

```diff
@@ -3986,28 +4176,53 @@ int clusterSendModuleMessageToTarget(const char *target,
  * Otherwise:
  * Publish this message across the slot (primary/replica).
  * -------------------------------------------------------------------------- */
-void clusterPropagatePublish(robj *channel, robj *message, int sharded) {
-    clusterMsgSendBlock *msgblock;
+void clusterPropagatePublish(robj *channel, robj **messages, int count, int sharded) {
+    /* Currently publish command does not support multiple payload. */
+    serverAssert(count == 1);
 
-    if (!sharded) {
-        msgblock = clusterCreatePublishMsgBlock(channel, message, CLUSTERMSG_TYPE_PUBLISH);
-        clusterBroadcastMessage(msgblock);
-        clusterMsgSendBlockDecrRefCount(msgblock);
-        return;
-    }
+    clusterMsgSendBlock *msgblock, *msgblock_light;
+    int nodes_not_supporting_light_header = 0;
 
-    listIter li;
-    listNode *ln;
-    list *nodes_for_slot = clusterGetNodesInMyShard(server.cluster->myself);
-    serverAssert(nodes_for_slot != NULL);
-    listRewind(nodes_for_slot, &li);
-    msgblock = clusterCreatePublishMsgBlock(channel, message, CLUSTERMSG_TYPE_PUBLISHSHARD);
-    while ((ln = listNext(&li))) {
-        clusterNode *node = listNodeValue(ln);
-        if (node->flags & (CLUSTER_NODE_MYSELF | CLUSTER_NODE_HANDSHAKE)) continue;
-        clusterSendMessage(node->link, msgblock);
+    msgblock_light = clusterCreatePublishLightMsgBlock(
+        channel, messages, count, sharded ? CLUSTERMSG_TYPE_PUBLISHSHARD_LIGHT : CLUSTERMSG_TYPE_PUBLISH_LIGHT);
+
+    if (server.cluster->all_nodes_are_known_to_support_light_hdr) {
+        /* If the cluster is homogeneous broadcast to all nodes */
+        if (sharded) {
+            clusterShardBroadcastMessage(msgblock_light);
+        } else {
+            clusterBroadcastMessage(msgblock_light);
+        }
+    } else {
```

Comment:
IMO this just adds extra complexity and duplicated logic. Broadcast is just a loop over the nodes anyway, so we can just do our own loop in all cases. We don't need `clusterShardBroadcastMessage` either.

The other benefits that we can remove `server.cluster->all_nodes_are_known_to_support_light_hdr` variable and the logic for keeping it updated. This is not a performance critical code path in any way, so it's better to keep it simple.

Lazy initialization of msgblock would be good though, to only create it when we see a node that needs it.